File: csharp/meetup/Meetup.cs
using System;

enum Schedule
{
  First,
  Second,
  Third,
  Fourth,
  Teenth,
  Last
}

class Meetup
{
  public Meetup(int aMonth, int aYear)
  {
    month = aMonth;
    year = aYear;
  }

  private DateTime GetNthOccuranceDayOfMonth(int nthOccurance, DayOfWeek dayOfWeek)
  {
    var occurance = 0;
    DateTime day = new DateTime();
    for (var i = 1; i <= DateTime.DaysInMonth(year, month); i++)
    {
      day = new DateTime(year, month, i);
      if (day.DayOfWeek == dayOfWeek)
      {
        occurance++;
        if (occurance == nthOccurance)
          break;
      }
    }
    return day;
  }

  public DateTime Day(DayOfWeek dayOfWeek, Schedule schedule)
  {
    DateTime day = new DateTime(year, month, 1);

    switch (schedule)
    {
      case Schedule.First:
        day = GetNthOccuranceDayOfMonth(1, dayOfWeek);
        break;
      case Schedule.Second:
        day = GetNthOccuranceDayOfMonth(2, dayOfWeek);
        break;
      case Schedule.Third:
        day = GetNthOccuranceDayOfMonth(3, dayOfWeek);
        break;
      case Schedule.Fourth:
        day = GetNthOccuranceDayOfMonth(4, dayOfWeek);
        break;
      case Schedule.Teenth:
        for (var i = 13; i <= 19; i++)
        {
          day = new DateTime(year, month, i);
          if (day.DayOfWeek == dayOfWeek)
            break;
        }
        break;
      case Schedule.Last:
        for (var i = DateTime.DaysInMonth(year, month); i >= 1; i--)
        {
          day = new DateTime(year, month, i);
          if (day.DayOfWeek == dayOfWeek)
            break;
        }
        break;
    }
    return day;
  }

  private int month;
  private int year;
}

File: javascript/meetup/meetup.js
var weekDays = "Sunday Monday Tuesday Wednesday Thursday Friday Saturday".split(" ");

var Meetup = function (year, month, day, nth) {
  var intDay = weekDays.indexOf(day);

  switch (nth) {
    case "teenth":
      return find_day(year, month, intDay, 13, 19);
    case "1st":
      return find_day(year, month, intDay, 1, 7);
    case "2nd":
      return find_day(year, month, intDay, 8, 14);
    case "3rd":
      return find_day(year, month, intDay, 15, 21);
    case "4th":
      return find_day(year, month, intDay, 22, 28);
    case "5th":
      return find_day(year, month, intDay, 29, 31);
    case "last":
      return last(year, month, intDay);
  }
};

function find_day(year, month, day, start, stop) {
  for (var meetup = new Date(year, month, start);
       meetup.getDate() <= stop && meetup.getMonth() == month;
       meetup = meetup.addDays(1)) {

    if (meetup.getDay() == day)
      return meetup;
  }
  throw "Date does not exist";
}

function last(year, month, day) {
  for (var meetup = new Date(year, month + 1, 0);
       meetup.getMonth() == month;
       meetup = meetup.addDays(-1)) {

    if (meetup.getDay() == day)
      return meetup;
  }
  throw "Date does not exist";
}

Date.prototype.addDays = function (days) {
  var dat = new Date(this.valueOf());
  dat.setDate(dat.getDate() + days);
  return dat;
};

module.exports = Meetup;